### DISCUSSION TITLE
Best practices: creating an extension or modifying a plugin

### DISCUSSION THREAD

**Discussion by amunn** [jspsych/jsPsych#1842] [Q&A]
For one of our experiments, we're displaying a picture along with an audio prompt. I've modified the `audio-button-response` plugin to accept an `image` parameter, but I'm wondering if this is the sort of thing that would be better done with an extension, so that it could be added to the other response types? If so, what would such an extension look like? I'm a bit lost on how they work altogether.

**Comment by jodeleeuw** [comment]
Hi @amunn,

Extensions are meant to be code that could be added to (nearly) any plugin and create features that "sit on top" of the plugin without interfering with the plugin. Eye tracking, mouse tracking, and key logging would all be good candidates for an extension. I've also been contemplating adding an extension that simply plays an audio file for cases where the exact timing of the audio isn't essential, e.g., background music during a task or audio instructions.

Extensions work by defining a set of functions that are called at particular times. For example, there's one for when a trial starts, and this can be used to do something extra on top of what the plugin does at the start of the trial, like start an eye tracker.

**Reply by amunn** [reply]
Thanks @jodeleeuw. So I assume for what I'm doing modifying the plugin is the appropriate choice, since the image I'm adding is part of the stimuli.
